training-material/checklist.md

```diff
@@ -0,0 +1,104 @@
+# A proof checklist for writers and reviewers
+
+This is a check list of best practices intended
+
+* for proof writers to use before checking a proof into a repository and
+* for proof reviewers to use during the code review of a pull
+  request containing a proof.
+
+The best practices for writing a proof are already described in detail
+in the section [How to Write a CBMC Proof](PROOF-WRITING.md)
+of the [training material](README.md)
+win this [starter kit](../README.md).
+
+There are two questions whose answers must be absolutely clear:
+
+* What properties are being checked by the proof?
+* What assumptions are being made by the proof?
+
+These answers should be found in one of three places:
+
+* The proof harness,
+* The proof makefiles (and, with the starter kit, these makefiles are
+  the proof Makefile, the project Makefile-project-defines,
+  and the project Makefile.common), and perhaps
+* The proof readme file.
+
+## Properties checked
+
+* Run cbmc with the following property-checking flags
+
+	* --malloc-may-fail
+	* --malloc-fail-null
+	* --bounds-check
+	* --conversion-check
+	* --div-by-zero-check
+	* --float-overflow-check
+	* --nan-check
+	* --pointer-check
+	* --pointer-overflow-check
+	* --pointer-primitive-check
+	* --signed-overflow-check
+	* --undefined-shift-check
+	* --unsigned-overflow-check
+
+  The starter kit uses these flags by default.  But the user of the starter kit
+  may disable any one of these flags by editing project Makefile.common or
+  by setting a makefile variable to the empty string
+  (as in `CBMC_FLAG_MALLOC_MAY_FAIL = `)
+  in the project Makefile-project-defines or a proof Makefile.
+
+* Document the decision to omit one of these property-checking flags.
+  There are valid reasons to omit them either for a project or for an
+  individual proof. But the decision and the reason for the decision
+  must be documented either in a project readme or a proof readme file.
+
+* CBMC checks assertions in the code.  This is understood and need not be
+  documented.
+
+## Assumptions made
+
```

Comment:
Another example: 
* For each non-trivial data structure, an [`ensure_allocated` function](PROOF-WRITING.md#the-ensure_allocated-function) as described in the training material, has been defined.

  Feel free to use any naming scheme that makes sense for your project --- some
  projects use `allocate_X` in place of `ensure_allocated_X` --- but be
  consistent.

Reply:
Done.  Thanks.